### PR TITLE
Allow contibuting additional global variables for skipif/xfail

### PR DESCRIPTION
**Pull request by s0undt3ch**
- [ ] Include documentation when adding new features.
- [x] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.
- [x] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/master/changelog/README.rst) for details.

### CODE REVIEW

**Comment by s0undt3ch** [comment]
Please advise on the hook name, what additional testing might be required, or anything else regarding the PR.

**Comment by RonnyPfannschmidt** [comment]
Something like `pytest_markeval_namespace`, it's tricky to name it well

**Comment by RonnyPfannschmidt** [comment]
Also thank you for digging into this,
That looks good and takes something off my minds backlog.

Now the truely hard part will be the correct name of the hook 😅

File: src/_pytest/skipping.py [hunk 1]
```diff
@@ -101,6 +101,10 @@ def evaluate_condition(item: Item, mark: Mark, condition: object) -> Tuple[bool,
             "platform": platform,
             "config": item.config,
         }
+        for dictionary in item.config.hook.pytest_skipif_additional_globals():
+            if not isinstance(dictionary, dict) or not dictionary:
+                continue
+            globals_.update(dictionary)
```
Comment (RonnyPfannschmidt) [inline src/_pytest/skipping.py:107]: There may be need for a warning in case of conflicts

Comment (s0undt3ch) [reply src/_pytest/skipping.py:107]: So we allow overriding the `globals_` keys, but we should warn, correct?
Is there a good example on how to raise that warning?

And to be clear, we should not error, just warn.

**Comment by bluetech** [comment]
I personally consider string conditions to be a legacy feature, kept for compatibility, and better replaced by regular boolean conditions. Can you describe your use case for this hook?

**Comment by RonnyPfannschmidt** [comment]
@bluetech i would like this to land even if strings eventually go "away", because they can be arguments to a lambda function

**Comment by bluetech** [comment]
Is there an example for a situation where this would be useful?

**Comment by RonnyPfannschmidt** [comment]
i have a direct use-case
that boils down to the following pseudo code

```python
import attr

@attr.s
class AppUnderTestPlugin:
    _app = attr.s()

   @classmethod
   def from_config(cls, pytestconfig):
      ...
      return cls(...)
   

   def pytest_markeval_namespace(self):
     return {'application': self.app}

   @pytest.fixture
  def app(self):
    return self._app

def pytest_configure(config):
     config.pluginmanager.register(AppUnderTestPlugin.from_config(config))

```

`test_fun.py`
```python
import pytest

@pytest.mark.xfail("application.use_beta", reason="only in stable"
def test_fun(app):
   ...
```

**Comment by s0undt3ch** [comment]
Our project, [salt](https://github.com/saltstack/salt), is a multiple platform project, and, as such, we have tests targeted to specific platforms, and in some cases, to specific versions of said platforms. It constructs a dictionary which we call [grains](https://docs.saltstack.com/en/latest/topics/grains/) that includes several details about the platform, consider it [distro](https://pypi.org/project/distro/) on some serious steroids abuse.

Constructing this grains dictionary is an expensive runtime operation, and for the most part, these _grains_ are static, so, constructing them on every module that needs to use them is unpractical. We do have the option to construct these grains in a global module variable and import it so that we don't have to use "string conditions", but that means a global mutable dictionary in a module scope and requiring it's import on every test module than needs to use them.

"String conditions" make this whole code a lot simpler, and it would be a pity to see if go away.

Example code:
```python
@pytest.mark.skipif(
    "grains['os_family'] == 'RedHat' and grains.get('osmajorrelease') == 6",
    reason='RedHat 6 Specific Test'
)
def test_foo():
    assert False
```

**Comment by s0undt3ch** [comment]
Hmm, why isn't [this line](https://codecov.io/gh/pytest-dev/pytest/compare/877c62166ac3557b424f4e1d08ceb31427ebf970...c66f12d764afd3819a24ef291f27d93fba244419/diff) getting hit by tests? Do hooks ignore empty results?

File: src/_pytest/skipping.py (hunk 1 above)
Comment (nicoddemus) [reply src/_pytest/skipping.py:107]: I would vote to do not warn or error out. Most users expect that something updating a dictionary will overwrite what's in there.

Let's see how this plays out first, adding a warning is simple to do later if we get reports of hard to tracking issues.

**Comment by nicoddemus** [comment]
> i have a direct use-case

I agree this is useful, because while this will be used only for string conditions (which I agree we might want to get away in the future, but not sure), the actual feature is "delayed" evaluation. As @RonnyPfannschmidt mentions, we can evolve this in the future to make the condition also be a callable which supports arbitrary keyword arguments. String conditions and callable conditions would then be evaluated at test setup time, and not during collection. 

Adapting @RonnyPfannschmidt's example:

```python
import pytest

@pytest.mark.xfail(lambda **kw: kw["application"].use_beta, reason="only in stable")
def test_fun(app):
   ...
```

Or @s0undt3ch's:

```python
@pytest.mark.skipif(
    lambda **kw: kw["grains"]['os_family'] == 'RedHat' and kw["grains"].get('osmajorrelease') == 6,
    reason='RedHat 6 Specific Test'
)
def test_foo():
    assert False
``` 

I've also used string evaluation in modern pytest for the same reason.

**Comment by RonnyPfannschmidt** [comment]
The lamdas could take the argument names directly, no need for kw

File: testing/test_skipping.py [hunk 2]
```diff
@@ -153,6 +154,71 @@ def test_func(self):
         assert skipped
         assert skipped.reason == "condition: config._hackxyz"
 
+    def test_skipif_additional_globals(self, testdir):
```
Comment (nicoddemus) [inline testing/test_skipping.py:157]: Those tests are great, but please also add a test where different conftests provide different values, and the right `pytest_markeval_namespace` hook wins based on the test location. Something like:

```
tests/conftest.py   # pytest_markeval_namespace() -> {"arg": "root"}
tests/foo/conftest.py   # pytest_markeval_namespace() -> {"arg": "foo"}
tests/bar/conftest.py   # pytest_markeval_namespace() -> {"arg": "bar"}

tests/test_root.py  # single test with skipif("arg == 'root')
tests/foo/test_foo.py  # single test with skipif("arg == 'foo')
tests/bar/test_bar.py  # single test with skipif("arg == 'bar')
```

The above setup should make all tests skip. With the current code only `test_root` will be skipped.

**Comment by nicoddemus** [comment]
> The lamdas could take the argument names directly, no need for kw

If we inspect the function to only pass the arguments it declares (like we do with hooks), you are right of course. 👍

File: src/_pytest/skipping.py [hunk 3]
```diff
@@ -101,6 +101,10 @@ def evaluate_condition(item: Item, mark: Mark, condition: object) -> Tuple[bool,
             "platform": platform,
             "config": item.config,
         }
+        for dictionary in item.config.hook.pytest_markeval_namespace():
+            if not isinstance(dictionary, dict) or not dictionary:
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply src/_pytest/skipping.py:105]: Makes sense, will update.

File: testing/test_skipping.py (hunk 2 above)
Comment (s0undt3ch) [reply testing/test_skipping.py:157]: I'll add your suggested scenario.

Comment (s0undt3ch) [reply testing/test_skipping.py:157]: I'll add your suggested scenario.

Comment (s0undt3ch) [reply testing/test_skipping.py:157]: I'll add your suggested scenario.

File: src/_pytest/skipping.py [hunk 4]
```diff
@@ -101,6 +101,10 @@ def evaluate_condition(item: Item, mark: Mark, condition: object) -> Tuple[bool,
             "platform": platform,
             "config": item.config,
         }
+        for dictionary in item.config.hook.pytest_markeval_namespace():
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply src/_pytest/skipping.py:104]: Thanks!
Did not know about the `item.ihook` and what that entails.
Will update.

File: src/_pytest/hookspec.py [hunk 5]
```diff
@@ -808,6 +808,16 @@ def pytest_warning_recorded(
     """
 
 
+# -------------------------------------------------------------------------
+# Hooks for influencing skipping
+# -------------------------------------------------------------------------
+def pytest_markeval_namespace() -> Dict[str, Any]:
+    """Called when constructing the globals dictionary used for
+    evaluating skipif contditions.
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply src/_pytest/hookspec.py:816]: :+1:

File: src/_pytest/skipping.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply src/_pytest/skipping.py:104]: Also, if we add `config` as an argument, I need to pass it on my call correct?
```python
for dictionary in item.ihook.pytest_markeval_namespace(item.config):
    ...
```

[the remark this answers is not in the document]
Comment (s0undt3ch) [reply src/_pytest/skipping.py:104]: Im still learning how much magic pytest does for us :)

File: testing/test_skipping.py [hunk 6]
```diff
@@ -153,6 +154,127 @@ def test_func(self):
...
+            def test_2():
+                assert True
+        """
+        )
+        res = testdir.runpytest(p)
+        assert res.ret == 0
+        res.stdout.fnmatch_lines(["*1 skipped*"])
+        res.stdout.fnmatch_lines(["*1 passed*"])
+
+    def test_skipif_markeval_namespace_multiple(self, testdir):
```
Comment (s0undt3ch) [inline testing/test_skipping.py:184]: @nicoddemus, I added the suggested scenario, however, root skips while the other two don't.

How are `ihooks` evaluated?
Top level conftest then lower level conftest? That's how I'm expecting it to work but the test scenario is proving me wrong...

File: changelog/7695.feature.rst [hunk 7]
```diff
@@ -0,0 +1,19 @@
+A new hook was added, `pytest_markeval_namespace` which should return a dictionary.
+This dictionary will be used to augment the "global" variables available to evaluate skipif/xfail/xpass markers.
+
+Psudo example
```
Comment (The-Compiler) [inline changelog/7695.feature.rst:4]:
```suggestion
Pseudo example
```

**Comment by s0undt3ch** [comment]
Ok.
After some crafty debug print statements,
```diff
diff --git a/src/_pytest/skipping.py b/src/_pytest/skipping.py
index 92916eb8b..dce3f90d8 100644
--- a/src/_pytest/skipping.py
+++ b/src/_pytest/skipping.py
@@ -103,6 +103,7 @@ def evaluate_condition(item: Item, mark: Mark, condition: object) -> Tuple[bool,
             "config": item.config,
         }
         for dictionary in item.ihook.pytest_markeval_namespace(config=item.config):
+            print(1, dictionary)
             if not isinstance(dictionary, Mapping):
                 raise ValueError(
                     "pytest_markeval_namespace() needs to return a dict, got {!r}".format(
@@ -110,6 +111,7 @@ def evaluate_condition(item: Item, mark: Mark, condition: object) -> Tuple[bool,
                     )
                 )
             globals_.update(dictionary)
+        print(2, globals_)
         if hasattr(item, "obj"):
             globals_.update(item.obj.__globals__)  # type: ignore[attr-defined]
         try:
diff --git a/testing/test_skipping.py b/testing/test_skipping.py
index 21a1bed66..c542e4703 100644
--- a/testing/test_skipping.py
+++ b/testing/test_skipping.py
@@ -190,6 +190,7 @@ class TestEvaluation:
             import pytest
 
             def pytest_markeval_namespace():
+                print("root")
                 return {"arg": "root"}
             """
             )
@@ -213,6 +214,7 @@ class TestEvaluation:
             import pytest
 
             def pytest_markeval_namespace():
+                print("foo")
                 return {"arg": "foo"}
             """
             )
@@ -236,6 +238,7 @@ class TestEvaluation:
             import pytest
 
             def pytest_markeval_namespace():
+                print("bar")
                 return {"arg": "bar"}
             """
             )
```
hooks are getting called nested first, top level last.
```
testing/test_skipping.py::TestEvaluation::test_skipif_markeval_namespace_multiple =============================================================================================== test session starts ==============
==================================================================================                                                                                                                                  
platform linux -- Python 3.7.5, pytest-6.0.1.dev154+gbf3dcce91.d20200830, py-1.9.0, pluggy-0.13.1 -- /home/vampas/projects/SaltStack/pytest/pytest/features/skip-context-hook/.tox/py37/bin/python                  
cachedir: .pytest_cache                              
rootdir: /tmp/pytest-of-vampas/pytest-27/test_skipif_markeval_namespace_multiple0
collected 3 items                                                                                                                                                                                                   
                                                     
root/test_root.py::test_root root                                                                         
1 {'arg': 'root'}                                    
2 {'os': <module 'os' from '/home/vampas/.dotfiles/.ext/pyenv/versions/3.7.5/lib/python3.7/os.py'>, 'sys': <module 'sys' (built-in)>, 'platform': <module 'platform' from '/home/vampas/.dotfiles/.ext/pyenv/versions/3.7.5/lib/python3.7/platform.py'>, 'config': <_pytest.config.Config object at 0x7fdf433ca850>, 'arg': 'root'}                                                                                                     
SKIPPED                                                                                                   
root/bar/test_bar.py::test_bar bar                                                                        
root                                        
1 {'arg': 'bar'}                        
1 {'arg': 'root'}                         
2 {'os': <module 'os' from '/home/vampas/.dotfiles/.ext/pyenv/versions/3.7.5/lib/python3.7/os.py'>, 'sys': <module 'sys' (built-in)>, 'platform': <module 'platform' from '/home/vampas/.dotfiles/.ext/pyenv/version
s/3.7.5/lib/python3.7/platform.py'>, 'config': <_pytest.config.Config object at 0x7fdf433ca850>, 'arg': 'root'}
FAILED                                                                                                                                                                                                              
root/foo/test_foo.py::test_foo foo                  
root                                                                                                      
1 {'arg': 'foo'}                                                                                                                                                                                                    
1 {'arg': 'root'}                                                                                                                                                                                                   
2 {'os': <module 'os' from '/home/vampas/.dotfiles/.ext/pyenv/versions/3.7.5/lib/python3.7/os.py'>, 'sys': <module 'sys' (built-in)>, 'platform': <module 'platform' from '/home/vampas/.dotfiles/.ext/pyenv/version
s/3.7.5/lib/python3.7/platform.py'>, 'config': <_pytest.config.Config object at 0x7fdf433ca850>, 'arg': 'root'}                                                                                                     
FAILED
```

Is this how hooks are supposed to be called? I was expecting top level first, nested last.
What't the approach to take here?

File: src/_pytest/skipping.py [hunk 8]
```diff
@@ -101,6 +102,14 @@ def evaluate_condition(item: Item, mark: Mark, condition: object) -> Tuple[bool,
             "platform": platform,
             "config": item.config,
         }
+        for dictionary in item.ihook.pytest_markeval_namespace(config=item.config):
```
Comment (s0undt3ch) [inline src/_pytest/skipping.py:105]:
```suggestion
        for dictionary in reversed(item.ihook.pytest_markeval_namespace(config=item.config)):
```
Is this an acceptable solution?

**Comment by nicoddemus** [comment]
> Is this how hooks are supposed to be called? I was expecting top level first, nested last.
What't the approach to take here?

Yeah that's the correct order: hooks are called first for the nested plugins, then root conftest files, then thid-party plugins, and then lastly the builtin plugins. The logic here is that this allows plugins and local files to override the builtin behaviors.

File: testing/test_skipping.py [hunk 9]
```diff
@@ -153,6 +154,133 @@ def test_func(self):
...
```
Comment (nicoddemus) [inline testing/test_skipping.py:184]: I suggest we add a docstring here:

```python
"""
Keys defined by ``pytest_markeval_namespace()`` in nested plugins override top-level ones.
"""
```

File: src/_pytest/hookspec.py [hunk 10]
```diff
@@ -808,6 +808,18 @@ def pytest_warning_recorded(
     """
 
 
+# -------------------------------------------------------------------------
+# Hooks for influencing skipping
+# -------------------------------------------------------------------------
+def pytest_markeval_namespace(config: "Config") -> Dict[str, Any]:
+    """Called when constructing the globals dictionary used for
+    evaluating string conditions in xfail/skipif markers.
```
Comment (nicoddemus) [inline src/_pytest/hookspec.py:816]:
```suggestion
    evaluating string conditions in xfail/skipif markers.
    
    This is useful when the condition for a marker requires 
    objects that are expensive or impossible to obtain during 
    collection time, which is required by normal boolean 
    conditions.
    
    .. versionadded:: 6.1
```

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @s0undt3ch, looks great to me!

I left minor adjustments, please take a look.

**Comment by s0undt3ch** [comment]
Review comments addressed

**Review by nicoddemus — APPROVED** [review]
LGTM, thanks @s0undt3ch!

**Comment by bluetech** [comment]
Maybe we can add the lambda already, and let it accept fixtures? I think usually such a global value would be useful as a fixture anyway (in @RonnyPfannschmidt's example it already is). (Would be interested to know if in @s0undt3ch's example the grain can be a session-scoped fixture or such).

I haven't studied how fixtures work in detail yet, so can't say how viable it is. But I imagine it would just be treated as regular fixtures requested by the item (during the setup/call runtest phase, i.e. not recreated if the item requests the fixture on its own as well).

**Comment by nicoddemus** [comment]
> Maybe we can add the lambda already, and let it accept fixtures? I think usually such a global value would be useful as a fixture anyway (in @RonnyPfannschmidt's example it already is). (Would be interested to know if in @s0undt3ch's example the grain can be a session-scoped fixture or such).

This is a good idea and I haven't thought of it. 👍 

However I think this should be done in a separate issue/PR, given that this PR is an improvement already and will also be compatible to whatever we plan to do next.

**Comment by s0undt3ch** [comment]
@bluetech, yes it can be a session scoped fixture.

**Comment by bluetech** [comment]
It would be suboptimal to have redundant methods for achieving the same thing, and I think it will be nicer to reuse an existing feature rather than adding a new hook.

I propose that, if @s0undt3ch is willing to delay this to 6.2 (~ a few months), I will try to find some time to try and implement the fixture approach (or another one). But if it's too much delay, or if I (or someone else) don't get to it in time for 6.2, then we will merge this PR.

**Comment by s0undt3ch** [comment]
Our complete switch to PyTest has also been delayed until at least September 16th, and right now we're not relying on this PR to make the switch(on 5.x I was actually replacing the MarkEvaluator with a patched one which injected these grains, but I reverted that since I want us to migrate to 6.x).

With that said, I don't want to _force_ this PR going in if it's going to be obsoleted pretty soon.

I still thing a string with code in it _looks_ better than a lambda(and shorter), but that's personal preference and outside of the scope of this decision, guess I need to better understand how the new approach will work with existing fixtures.

Anyway, I'm OK with postponing this PR until 6.2 or eventually closing it if the newer approach lands by then.

**Comment by nicoddemus** [comment]
Thanks @s0undt3ch!

I've moved this PR to the 6.2 milestone then.

@bluetech could you please open a new issue outlining the new proposal, and link to this one?

**Comment by bluetech** [comment]
Opened issue #7712.

File: src/_pytest/hookspec.py [hunk 11]
```diff
@@ -808,6 +808,25 @@ def pytest_warning_recorded(
...
+
+    This is useful when the condition for a marker requires
+    objects that are expensive or impossible to obtain during
+    collection time, which is required by normal boolean
+    conditions.
+
+    .. versionadded:: 6.1
```
Comment (nicoddemus) [inline src/_pytest/hookspec.py:823]:
```suggestion
    .. versionadded:: 6.2
```

**Comment by nicoddemus** [comment]
This can now be merged given we've released 6.1 this week. 🎉 

@s0undt3ch do yout want to rebase and do the honors of squash/merging this? 😁 

Also if anybody wants to review this now would be the time to speak up. 👍

**Comment by s0undt3ch** [comment]
Sure!
But, we're not waiting for @bluetech's fixture approach?

**Comment by nicoddemus** [comment]
> But, we're not waiting for @bluetech's fixture approach?

Ahh sorry, yeah, I was reviewing 6.2's milestones and saw this, and forgot about the details. 🤦

**Comment by s0undt3ch** [comment]
Anyway, squashed and rebased.

**Comment by nicoddemus** [comment]
Thanks!

**Pull request closed without merging**